ios/RNMessageStream.m

```diff
@@ -40,14 +42,50 @@ - (instancetype)initWithDisplayInAppNotifications:(BOOL)displayNotifications {
 }
 
 - (BOOL)shouldPresentInAppNotificationForMessage:(MARMessage *)message {
-    NSMutableDictionary *payload = [NSMutableDictionary dictionaryWithDictionary:[message dictionary]];
-
-    if ([message attributes]) {
-        [payload setObject:[message attributes] forKey:@"attributes"];
+    if (self.defaultInAppNotification) {
+        return YES;
     }
+    __block BOOL result = YES;
+
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        result = [self emitWithTimeout:message];
+
+        dispatch_semaphore_signal(semaphore);
+    });
 
-    [self sendEventWithName:@"inappnotification" body:payload];
-    return self.displayInAppNotifications;
+    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
+
+    return result;
+}
+
+- (BOOL)emitWithTimeout:(MARMessage *)message {
+    __block BOOL success = YES;
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        NSMutableDictionary *payload = [NSMutableDictionary dictionaryWithDictionary:[message dictionary]];
+
+        if ([message attributes]) {
+            [payload setObject:[message attributes] forKey:@"attributes"];
+        }
+        [self sendEventWithName:@"inappnotification" body:payload];
+        
+        success = NO;
+        dispatch_semaphore_signal(self.eventSemaphore);
+    });
+
+    dispatch_semaphore_wait(self.eventSemaphore, dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC));
+
+    return success;
+}
+
+RCT_EXPORT_METHOD(acknowledgeEvent) {
```

Comment:
Would like it if we could pass a `BOOL` back here and use that as the return value. This would allow the RN layer to add logic around whether or not it wants to handle a given message then return the response.